fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,467 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::cache::SequentialKeyCache;
+use crate::error::Error;
+use crate::{
+    BlockHash, BlockHeight, Config, IPCParentFinality, ParentFinalityProvider, ParentViewProvider,
+};
+use async_stm::{abort, StmDynResult, StmResult, TVar};
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::ValidatorSet;
+
+type ParentViewPayload = (BlockHash, ValidatorSet, Vec<CrossMsg>);
+
+/// The default parent finality provider
+pub struct DefaultFinalityProvider {
+    config: Config,
+    parent_view_data: ParentViewData,
+    /// This is a in memory view of the committed parent finality,
+    /// it should be synced with the store committed finality, owner of the struct should enforce
+    /// this.
+    last_committed_finality: TVar<IPCParentFinality>,
+}
+
+/// Tracks the data from the parent
+#[derive(Clone)]
+struct ParentViewData {
+    height_data: TVar<SequentialKeyCache<BlockHeight, ParentViewPayload>>,
+}
+
+impl ParentViewData {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>> {
+        let cache = self.height_data.read()?;
+        // safe to unwrap, we dont allow no upper bound
```

Comment:
This comment is a bit confusing with the double negative, would you mind elaborating a bit? Thanks

Reply:
I removed this outdated comment, nice catch!